rust/lance-index/src/scalar/inverted/index.rs

```diff
@@ -993,6 +1007,94 @@ fn do_flat_full_text_search<Offset: OffsetSizeTrait>(
     Ok(results)
 }
 
+pub fn flat_bm25_search(
+    batch: RecordBatch,
+    doc_col: &str,
+    inverted_list: &InvertedListReader,
+    query_tokens: &HashMap<String, Option<u32>>,
+    tokenizer: &mut tantivy::tokenizer::TextAnalyzer,
+    avgdl: f32,
+    num_docs: usize,
+) -> std::result::Result<RecordBatch, DataFusionError> {
+    let doc_iter = iter_str_array(&batch[&doc_col]);
+    let mut scores = Vec::with_capacity(batch.num_rows());
+    for doc in doc_iter {
+        let doc = match doc {
+            Some(doc) => doc,
+            None => {
+                scores.push(0.0);
+                continue;
+            }
+        };
+        let doc_tokens = collect_tokens(doc, tokenizer);
+        let doc_norm = K1 * (1.0 - B + B * doc_tokens.len() as f32 / avgdl);
+        let mut doc_token_count = HashMap::new();
+        for token in doc_tokens {
+            doc_token_count
+                .entry(token)
+                .and_modify(|count| *count += 1)
+                .or_insert(1);
+        }
```

Comment:
perf: It seems like we are collecting all tokens, not just the ones relevant for the query. IIUC the only reason we are doing that is to get the total count of tokens. Maybe `collect_tokens` should take a parameter for tokens to filter for, and only save those tokens, but also return the total token count?

Reply:
Good point, fixed!